src/libcore/mem/manually_drop.rs

```diff
@@ -7,14 +7,20 @@ use crate::ptr;
 ///
 /// `ManuallyDrop<T>` is subject to the same layout optimizations as `T`.
 /// As a consequence, it has *no effect* on the assumptions that the compiler makes
-/// about all values being initialized at their type.  In particular, initializing
-/// a `ManuallyDrop<&mut T>` with [`mem::zeroed`] is undefined behavior.
+/// about its contents: the validity rules for `ManuallyDrop<T>` are the same as those
```

Comment:
Hm... in the `MaybeUninit` docs, we call these rules "initialization invariant". The term "validity invariant" is not official docs-level terminology, because many people are (justly) worried that "validity" is just way too ambiguous. That is, I think, why it referred to "initialized" here.

Reply:
Hm, IMO "initialisation invariant" is quite confusing in this context, for two reasons:

1) We use "uninitialised" to mean a specific state of memory, which is a different state from "not meeting the initialisation invariant". For example, for `MaybeUninit`, "uninitialised" *does* meet the initialisation invariant.

2) "Initialised" normally means the state between the value being constructed and its destruction. However, for `ManuallyDrop` we have the constraint that the contained value must continue to meet the initialisation invariant... after it is dropped.

If "initialisation invariant" is the terminology we have settled on for now, then I think we need to call out the above ambiguities.

Reply:
I agree that "initialization invariant" is not a good term for the reasons you gave.

> However, for ManuallyDrop we have the constraint that the contained value must continue to meet the initialisation invariant... after it is dropped.

Not really -- after `mem::drop::<ManuallyDrop<T>>(x)`, `x` is gone and its initialization does not matter any more.

The inherent method in `ManuallyDrop` does not drop the `ManuallyDrop`, as far as the language definition is concerned.